File: Samples/SocketSample.Android/app/src/main/java/com/neodamul/socketsample/ProtoCodec.java
package com.neodamul.socketsample;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public final class ProtoCodec {
    private ProtoCodec() {
    }

    public static byte[] clientRegister(long clientId) {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        writeVarintField(output, 1, clientId);
        writeVarintField(output, 2, System.currentTimeMillis());
        return output.toByteArray();
    }

    public static byte[] clientMessageSend(long sourceClientId, long targetClientId, String content) {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        writeStringField(output, 1, UUID.randomUUID().toString().replace("-", ""));
        writeVarintField(output, 2, sourceClientId);
        writeVarintField(output, 3, targetClientId);
        writeStringField(output, 4, content == null ? "" : content);
        writeVarintField(output, 5, 10);
        writeVarintField(output, 6, System.currentTimeMillis());
        return output.toByteArray();
    }

    public static boolean decodeRegisterAck(byte[] payload) {
        return parse(payload).bools.getOrDefault(2, false);
    }

    public static boolean decodeAckDelivered(byte[] payload) {
        return parse(payload).bools.getOrDefault(4, false);
    }

    public static String decodeErrorMessage(byte[] payload) {
        return parse(payload).strings.getOrDefault(5, "Message failed.");
    }

    public static ClientDelivery decodeDelivery(byte[] payload) {
        Parsed parsed = parse(payload);
        long source = parsed.varints.getOrDefault(2, 0L);
        long target = parsed.varints.getOrDefault(3, 0L);
        String content = parsed.strings.getOrDefault(4, "");
        return new ClientDelivery(source, target, content);
    }

    private static Parsed parse(byte[] payload) {
        Parsed parsed = new Parsed();
        int[] offset = new int[] { 0 };
        while (offset[0] < payload.length) {
            long key = readVarint(payload, offset);
            int field = (int)(key >> 3);
            int wireType = (int)(key & 0x7);
            if (wireType == 0) {
                long value = readVarint(payload, offset);
                parsed.varints.put(field, value);
                parsed.bools.put(field, value != 0);
            } else if (wireType == 2) {
                int length = (int)readVarint(payload, offset);
                if (offset[0] + length > payload.length) {
                    return parsed;
                }

                parsed.strings.put(field, new String(payload, offset[0], length, StandardCharsets.UTF_8));
                offset[0] += length;
            } else {
                return parsed;
            }
        }

        return parsed;
    }

    private static void writeVarintField(ByteArrayOutputStream output, int field, long value) {
        writeVarint(output, (field << 3));
        writeVarint(output, value);
    }

    private static void writeStringField(ByteArrayOutputStream output, int field, String value) {
        byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
        writeVarint(output, (field << 3) | 2);
        writeVarint(output, bytes.length);
        output.writeBytes(bytes);
    }

    private static void writeVarint(ByteArrayOutputStream output, long value) {
        long remaining = value;
        while (remaining >= 0x80) {
            output.write((int)((remaining & 0x7f) | 0x80));
            remaining >>= 7;
        }

        output.write((int)remaining);
    }

    private static long readVarint(byte[] data, int[] offset) {
        int shift = 0;
        long result = 0;
        while (offset[0] < data.length && shift <= 63) {
            int value = data[offset[0]++] & 0xff;
            result |= (long)(value & 0x7f) << shift;
            if ((value & 0x80) == 0) {
                return result;
            }

            shift += 7;
        }

        return result;
    }

    private static final class Parsed {
        final Map<Integer, Long> varints = new HashMap<>();
        final Map<Integer, Boolean> bools = new HashMap<>();
        final Map<Integer, String> strings = new HashMap<>();
    }

    public static final class ClientDelivery {
        public final long sourceClientId;
        public final long targetClientId;
        public final String content;

        ClientDelivery(long sourceClientId, long targetClientId, String content) {
            this.sourceClientId = sourceClientId;
            this.targetClientId = targetClientId;
            this.content = content;
        }
    }
}

File: Samples/SocketSample.AppleShared/ProtoCodec.swift
import Foundation

struct ClientMessageDelivery {
    let sourceClientId: UInt32
    let targetClientId: UInt32
    let content: String
}

struct ClientMessageAckResult {
    let targetClientId: UInt32
    let delivered: Bool
}

struct RouteTarget {
    let success: Bool
    let host: String
    let port: UInt16
    let errorMessage: String
}

enum ProtoCodec {
    static func routeRequest(clientId: UInt32) -> Data {
        var data = Data()
        data.appendVarintField(1, UInt64(clientId))
        data.appendStringField(3, "MostAvailableConnections")
        return data
    }

    static func decodeRouteResponse(_ data: Data) -> RouteTarget {
        let parsed = fields(from: data)
        return RouteTarget(
            success: parsed.bools[1] ?? false,
            host: parsed.strings[5] ?? "",
            port: UInt16(parsed.varints[6] ?? 0),
            errorMessage: parsed.strings[8] ?? "Route failed.")
    }

    static func clientRegister(clientId: UInt32) -> Data {
        var data = Data()
        data.appendVarintField(1, UInt64(clientId))
        data.appendVarintField(2, UInt64(Date().unixMilliseconds))
        return data
    }

    static func clientMessageSend(sourceClientId: UInt32, targetClientId: UInt32, content: String) -> Data {
        var data = Data()
        data.appendStringField(1, UUID().uuidString.replacingOccurrences(of: "-", with: ""))
        data.appendVarintField(2, UInt64(sourceClientId))
        data.appendVarintField(3, UInt64(targetClientId))
        data.appendStringField(4, content)
        data.appendVarintField(5, 10)
        data.appendVarintField(6, UInt64(Date().unixMilliseconds))
        return data
    }

    static func decodeRegisterAck(_ data: Data) -> Bool {
        fields(from: data).bools[2] ?? false
    }

    static func decodeDelivery(_ data: Data) -> ClientMessageDelivery? {
        let parsed = fields(from: data)
        guard let source = parsed.varints[2],
              let target = parsed.varints[3],
              let content = parsed.strings[4] else {
            return nil
        }

        return ClientMessageDelivery(
            sourceClientId: UInt32(source),
            targetClientId: UInt32(target),
            content: content)
    }

    static func decodeAck(_ data: Data) -> ClientMessageAckResult {
        let parsed = fields(from: data)
        return ClientMessageAckResult(
            targetClientId: UInt32(parsed.varints[3] ?? 0),
            delivered: parsed.bools[4] ?? false)
    }

    static func decodeErrorMessage(_ data: Data) -> String {
        let parsed = fields(from: data)
        return parsed.strings[5] ?? "Message failed."
    }

    private static func fields(from data: Data) -> ParsedFields {
        var offset = 0
        var parsed = ParsedFields()
        while offset < data.count {
            guard let key = data.readVarint(offset: &offset) else {
                break
            }

            let field = Int(key >> 3)
            let wireType = Int(key & 0x7)
            switch wireType {
            case 0:
                guard let value = data.readVarint(offset: &offset) else { return parsed }
                parsed.varints[field] = value
                parsed.bools[field] = value != 0
            case 2:
                guard let length = data.readVarint(offset: &offset) else { return parsed }
                let end = offset + Int(length)
                guard end <= data.count else { return parsed }
                parsed.strings[field] = String(data: data[offset..<end], encoding: .utf8) ?? ""
                offset = end
            default:
                return parsed
            }
        }

        return parsed
    }
}

private struct ParsedFields {
    var varints: [Int: UInt64] = [:]
    var bools: [Int: Bool] = [:]
    var strings: [Int: String] = [:]
}

private extension Data {
    mutating func appendFieldKey(_ field: Int, wireType: Int) {
        appendVarint(UInt64((field << 3) | wireType))
    }

    mutating func appendVarintField(_ field: Int, _ value: UInt64) {
        appendFieldKey(field, wireType: 0)
        appendVarint(value)
    }

    mutating func appendStringField(_ field: Int, _ value: String) {
        let bytes = Data(value.utf8)
        appendFieldKey(field, wireType: 2)
        appendVarint(UInt64(bytes.count))
        append(bytes)
    }

    mutating func appendVarint(_ value: UInt64) {
        var remaining = value
        while remaining >= 0x80 {
            append(UInt8(remaining & 0x7f) | 0x80)
            remaining >>= 7
        }

        append(UInt8(remaining))
    }

    func readVarint(offset: inout Int) -> UInt64? {
        var shift = 0
        var result: UInt64 = 0
        while offset < count && shift <= 63 {
            let byte = self[offset]
            offset += 1
            result |= UInt64(byte & 0x7f) << shift
            if (byte & 0x80) == 0 {
                return result
            }

            shift += 7
        }

        return nil
    }
}

private extension Date {
    var unixMilliseconds: Int64 {
        Int64(timeIntervalSince1970 * 1000)
    }
}
